daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/NormalizeRollbacksSpec.scala

```diff
@@ -0,0 +1,291 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf
+package speedy
+
+import org.scalatest._
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AnyWordSpec
+
+import com.daml.lf.data.ImmArray
+import com.daml.lf.data.Ref
+import com.daml.lf.transaction.Node.{GenNode, NodeRollback, NodeCreate, NodeExercises}
+import com.daml.lf.transaction.Transaction.LeafNode
+import com.daml.lf.transaction.TransactionVersion
+import com.daml.lf.transaction.{NodeId, GenTransaction}
+import com.daml.lf.value.{Value => V}
+
+class NormalizeRollbacksSpec extends AnyWordSpec with Matchers with Inside {
+
+  import NormalizeRollbackSpec._
+
+  // TODO: https://github.com/digital-asset/daml/issues/8020
+  // Below we test a set of hand-constructed testcases. Hopefully we got all edge cases!
+  // But for more confidence we might consider adding scalacheck based testing.
+
+  // TODO: https://github.com/digital-asset/daml/issues/8020
+  // We should test that the `meaning` of a transaction is preserved by normalization.
```

Comment:
Good idea. "Meaning" here would be the transaction tree among actions (i.e. after removing/flattening all rollback nodes), as well as the consistency order and the set of rolled back actions. That's a lot, but we definitely want to guarantee we're preserving these.